### PR TITLE
Document utils for functions and types

### PR DESCRIPTION
**Pull request by xuebinsu**
This patch documents the usage and notes when using the utils for
creating and using functions and types. It also renames some
possibly confuing concepts, including

- `as_name` to `column_name` to avoid the concept of "alias"
- `ArrayFunction` to `ColumnFunction` to reflect its use case

It also does other minor renaming and checks to improve
consistency of the interface.

### CODE REVIEW

**Review by yihong0618 — APPROVED** [review]
LGTM

FYI seems we did not change the `change` in `doc/source/notebooks/basic.ipynb`
code in this doc seems still the old one

**Comment by yihong0618** [comment]
LGTM~

File: greenplumpython/order.py [hunk 1]
```diff
@@ -80,7 +79,7 @@ def __getitem__(self, rows: slice) -> "DataFrame":
         Example:
             .. code-block::  Python
 
-                t.order_by("id").head(5)
+                t.order_by("id")[:5]
```
Comment (Quasimewdo) [inline greenplumpython/order.py:82]: I am thinking to open a PR to change all `t` in example to `df`

File: tests/test_func.py [hunk 2]
```diff
@@ -657,7 +659,7 @@ class sum_count_type(SimpleNamespace):
     def sum_count(result: sum_count_type, val: int) -> sum_count_type:
         if result is None:
             return SimpleNamespace(sum=val, count=1)
-        return SimpleNamespace(sum=result["sum"] + val, count=result["count"] + 1)
+        return SimpleNamespace(sum=result["count"] + val, count=result["count"] + 1)
```
Comment (Quasimewdo) [inline tests/test_func.py:662]: Why `sum=result["count"] + val`?

**Review by Quasimewdo — APPROVED** [review]
LGTM! Very informative

**Review by yihong0618 — APPROVED** [review]

**Pull request merged**